SimPEG/VRM/ProblemVRM.py

```diff
@@ -0,0 +1,609 @@
+            print("Refinement factor no longer matches length of refinement radii array. Please ensure refinement factor is equal or less to number of elements in refinement radii")
+
+        if Val > 4:
+            print("Refinement factor larger than 4 may result in computations which exceed memory limits")
+
+        self._refFact = Val
+
+    @property
+    def refRadius(self):
+        return self._refRadius
+
+    @refRadius.setter
+    def refRadius(self, Array):
+        assert isinstance(Array, np.ndarray), "Array must be a numpy array"
+
+        if self._refFact != len(Array):
+            print("Refinement factor no longer matches length of refinement radii array. Please ensure that the number of elements in refinement radii is equal or greater than the refinement factor")
+
+        self._refRadius = Array
+
+    @property
+    def topoMap(self):
+        return self._topoMap
+
+    @topoMap.setter
+    def topoMap(self, mappingObj):
+
+        assert isinstance(mappingObj, Maps.InjectActiveCells), "topoMap must be an instance of class Maps.InjectActiveCells"
+
+        self._topoMap = mappingObj
+        self._updateA = True
+
+    def _getH0matrix(self, xyz, pp):
+
+        """
+Creates sparse matrix containing inducing field components for source pp
+
+INPUTS:
+
+xyz: N X 3 array of locations to predict field
+pp: Source index
+        """
+
+        SrcObj = self.survey.srcList[pp]
+
+        H0 = SrcObj.getH0(xyz)
+
+        Hx0 = sp.diags(H0[:, 0], format="csr")
+        Hy0 = sp.diags(H0[:, 1], format="csr")
+        Hz0 = sp.diags(H0[:, 2], format="csr")
+
+        H0 = sp.vstack([Hx0, Hy0, Hz0])
+
+        return H0
+
+    def _getGeometryMatrix(self, xyzc, xyzh, pp):
+
+        """
+Creates the dense geometry matrix mapping from magentized voxel cells to the
+receivers for source pp
+
+INPUTS:
+
+xyzc: N by 3 array containing cell center locations
+xyzh: N by 3 array containing cell dimensions
+pp: Source index
+        """
+
+        srcObj = self.survey.srcList[pp]
+
+        N = np.shape(xyzc)[0]   # Number of cells
+        K = srcObj.nRx          # Number of receiver in all rxList
+
+        ax = np.reshape(xyzc[:, 0] - xyzh[:, 0]/2, (1, N))
+        bx = np.reshape(xyzc[:, 0] + xyzh[:, 0]/2, (1, N))
+        ay = np.reshape(xyzc[:, 1] - xyzh[:, 1]/2, (1, N))
+        by = np.reshape(xyzc[:, 1] + xyzh[:, 1]/2, (1, N))
+        az = np.reshape(xyzc[:, 2] - xyzh[:, 2]/2, (1, N))
+        bz = np.reshape(xyzc[:, 2] + xyzh[:, 2]/2, (1, N))
+
+        G = np.zeros((K, 3*N))
+        C = -(1/(4*np.pi))
+        eps = 1e-10
+
+        COUNT = 0
+
+        for qq in range(0, len(srcObj.rxList)):
+
+            rxObj = srcObj.rxList[qq]
+            dComp = rxObj.fieldComp
+            locs = rxObj.locs
+            M = np.shape(locs)[0]
+
+            if dComp is 'x':
+                for rr in range(0, M):
+                    u1 = locs[rr, 0] - ax
+                    u1[np.abs(u1) < 1e-10] = np.min(xyzh[:, 0])/1000
```

Comment:
- instead of hard-coding 1e-10, could you use the variable `eps`?
- what is the factor of 1000 for? is it a unit conversion?

Reply:
Values here were determine experimentally and provide stability to the forward model. If this is not done, then there may be locations relative to the cells in which we wind up dividing by zero.

Reply:
This was done